### ISSUE TITLE
self.ema.ema_model.sample(batch_size=n)  Can only generate noisy pictures

### ISSUE THREAD

**Issue by Maydaybye** [lucidrains/denoising-diffusion-pytorch#230]
I saw similar questions and solutions in the earlier #134 
But the author changed the code and the previous solution failed.

![image](https://github.com/lucidrains/denoising-diffusion-pytorch/assets/75834797/d45b26e4-eeff-4f1d-bb4f-c3f973a36d78)
Is this pytorch implementation still working?

**Comment by lucidrains** [comment]
@Maydaybye what dataset is that? maybe i'll just do some training and debug what is going on

it looks like training collapsed somewhere around 6k

**Comment by lucidrains** [comment]
could you share your training curves?

**Comment by Maydaybye** [comment]
So glad receving your reply!!
I'm using code version 1.8.3. 
The above result is trained on a portrait image from celebA.But I didn't save the training curve.
I've tried several times with the similar code below on different number of potrait images.Losses is between 0.8 and 0.01, but only noise images can be generated.
`import torch
from torchvision import transforms as T, utils
from denoising_diffusion_pytorch import Unet, GaussianDiffusion, Trainer

device = torch.device('cuda:' + str(2))
model = Unet(
    dim=64,
    dim_mults=(1, 2, 4, 8),
    flash_attn=True
)
model.to(device)
diffusion = GaussianDiffusion(
    model,
    image_size=128,
    timesteps=1000,  # number of steps
    sampling_timesteps=250
    # number of sampling timesteps (using ddim for faster inference [see citation for ddim paper])
)
diffusion.to(device)
trainer = Trainer(
    diffusion,
    'cceleba_1-100',
    train_batch_size=1,
    train_lr=8e-5,
    train_num_steps=700000,  # total training steps
    gradient_accumulate_every=2,  # gradient accumulation steps
    ema_decay=0.995,  # exponential moving average decay
    amp=True,  # turn on mixed precision
    calculate_fid=False  # whether to calculate fid during training
)

trainer.train()`

**Comment by Maydaybye** [comment]
> @Maydaybye what dataset is that? maybe i'll just do some training and debug what is going on
> 
> it looks like training collapsed somewhere around 6k

So glad receving your reply!!
I'm using code version 1.8.3.
The above result is trained on a portrait image from celebA.But I didn't save the training curve.
I've tried several times with the similar code below on different number of potrait images.Losses is between 0.8 and 0.01, but only noise images can be generated.
`import torch
from torchvision import transforms as T, utils
from denoising_diffusion_pytorch import Unet, GaussianDiffusion, Trainer

device = torch.device('cuda:' + str(2))
model = Unet(
dim=64,
dim_mults=(1, 2, 4, 8),
flash_attn=True
)
model.to(device)
diffusion = GaussianDiffusion(
model,
image_size=128,
timesteps=1000, # number of steps
sampling_timesteps=250
# number of sampling timesteps (using ddim for faster inference [see citation for ddim paper])
)
diffusion.to(device)
trainer = Trainer(
diffusion,
'cceleba_1-100',
train_batch_size=1,
train_lr=8e-5,
train_num_steps=700000, # total training steps
gradient_accumulate_every=2, # gradient accumulation steps
ema_decay=0.995, # exponential moving average decay
amp=True, # turn on mixed precision
calculate_fid=False # whether to calculate fid during training
)

trainer.train()`

**Comment by Maydaybye** [comment]
> @Maydaybye what dataset is that? maybe i'll just do some training and debug what is going on
> 
> it looks like training collapsed somewhere around 6k

So glad receving your reply!!
I'm using code version 1.8.3.
The above result is trained on a portrait image from celebA.But I didn't save the training curve.
I've tried several times with the similar code below on different number of potrait images.Losses is between 0.8 and 0.01, but only noise images can be generated.
`import torch
from torchvision import transforms as T, utils
from denoising_diffusion_pytorch import Unet, GaussianDiffusion, Trainer

device = torch.device('cuda:' + str(2))
model = Unet(
    dim=64,
    dim_mults=(1, 2, 4, 8),
    flash_attn=True
)
model.to(device)
diffusion = GaussianDiffusion(
    model,
    image_size=128,
    timesteps=1000,  # number of steps
    sampling_timesteps=250
    # number of sampling timesteps (using ddim for faster inference [see citation for ddim paper])
)
diffusion.to(device)
trainer = Trainer(
    diffusion,
    'cceleba_1-100',
    train_batch_size=1,
    train_lr=8e-5,
    train_num_steps=700000,  # total training steps
    gradient_accumulate_every=2,  # gradient accumulation steps
    ema_decay=0.995,  # exponential moving average decay
    amp=True,  # turn on mixed precision
    calculate_fid=False  # whether to calculate fid during training
)

trainer.train()`

**Comment by lucidrains** [comment]
@Maydaybye so before i give it a shot, do you want to try setting the `train_batch_size = 4`, the `gradient_accumulate_every = 8`? your effective batch size is 2, which is way too small

**Comment by Maydaybye** [comment]
> @Maydaybye so before i give it a shot, do you want to try setting the `train_batch_size = 4` and the `gradient_accumulate_every = 8`? your effective batch size is 2, which is way too small

Thank you very much!! I'll try it now.

**Comment by lucidrains** [comment]
@Maydaybye how did it go?

**Comment by Maydaybye** [comment]
> @Maydaybye how did it go?

I've tried with the code below.But still, only random pictures can be generated
Still using a picture as input.

training picture:
![000001](https://github.com/lucidrains/denoising-diffusion-pytorch/assets/75834797/59790882-449e-4ad1-b8fa-44d3161c0aa7)

sampling:
![image](https://github.com/lucidrains/denoising-diffusion-pytorch/assets/75834797/3e14250c-bb4a-4c18-8a75-7723e0e59905)

code:
`device = torch.device('cuda:' + str(0))
model = Unet(
    dim=64,
    dim_mults=(1, 2, 4, 8),
    flash_attn=True
)
model.to(device)
diffusion = GaussianDiffusion(
    model,
    image_size=128,
    timesteps=1000,  # number of steps
    sampling_timesteps=250
    # number of sampling timesteps (using ddim for faster inference [see citation for ddim paper])
)
diffusion.to(device)
trainer = Trainer(
    diffusion,
    'cceleba_1-100',
    save_and_sample_every=10,
    train_batch_size=4,
    train_lr=8e-5,
    train_num_steps=700000,  # total training steps
    gradient_accumulate_every=8,  # gradient accumulation steps
    ema_decay=0.995,  # exponential moving average decay
    amp=True,  # turn on mixed precision
    calculate_fid=False  # whether to calculate fid during training
)

trainer.train()`

**Comment by lucidrains** [comment]
how many images do you have in your `cceleba_1-100` folder? greater than 100 i hope

**Comment by Maydaybye** [comment]
Only one image.
![000001](https://github.com/lucidrains/denoising-diffusion-pytorch/assets/75834797/f0225d3f-1126-4b0c-a169-f27ceb4f6e4b)
Can't ddpm get good results on one image?

**Comment by lucidrains** [comment]
no, you need at least 10k images on the low end, even with augmentation

**Comment by lucidrains** [comment]
there is a paper doing what you describe https://arxiv.org/abs/1905.01164 , wonder if anyone has made it work with ddpms

**Comment by Maydaybye** [comment]
> there is a paper doing what you describe https://arxiv.org/abs/1905.01164 from GAN literature; wonder if anyone has made it work with ddpms. however, this repo is not setup to do single image training as in that paper

I'll try again with more than 10k images. Really appreciate your help and the relative paper!!! This gives me the motivatio!

**Comment by lucidrains** [comment]
you should just do the entire celeb dataset, which has 200k faces I believe

**Comment by Maydaybye** [comment]
> you should just do the entire celeb dataset, which has 200k faces I believe
Yes!! There is 202599 images in celeb datasets. 'm working on it now. ٩(^㉨^*)و✧

**Comment by Maydaybye** [comment]
> you should just do the entire celeb dataset, which has 200k faces I believe

Yes!! There is 202599 images in celeb datasets. 'm working on it now. ٩(^㉨^*)و✧

**Comment by Maydaybye** [comment]
It works with 202599 images in celeba datasets.
![image](https://github.com/lucidrains/denoising-diffusion-pytorch/assets/75834797/2aee12f1-e347-4d43-b740-52d7196be4e6)
These are the images smpled in the first 4000 iterations
ddpm has certain requirements on the number of training images to learn sufficient features.

**Comment by lucidrains** [comment]
nice! yea I'll add a few assertions for batch size and number of images in dataset tomorrow
